select nodes by xversion features

gandrewstone opened the pull request:
and use different hostnames to choose your selection.

This entire repo is not pretty code, its a large copy-paste from ancient bitcoind source code.  This PR makes no attempt to fix this problem.

Rather than working to make this code pretty, we should reuse the bitcoind source by pulling the dns code into our bitcoind repo and creating bu_dns executable as part of our release.

This would be a lot of work though -- almost as much as making this code pretty.  

Instead, let's get the job done.

gandrewstone committed:
select nodes by xversion features, and use different hostnames to choose your selection

sickpig commented on db.h:
```diff
@@ -373,16 +383,22 @@ class CAddrDb
+    {
+        CRITICAL_BLOCK(cs)
+        for (int i = 0; i < vAddr.size(); i++)
+            Add_(vAddr[i], CServiceResult(), fForce);
```
where `CServiceResult()` is defined? I see the struct defined in `db.h` but no function with that name.

gandrewstone replied:
It executes the default constructor

sickpig replied:
Ouch.Dumb question, isn't it? :smile:

sickpig commented on xversionmessage.cpp:
```diff
@@ -0,0 +1,62 @@
+//#include "random.h"
+//#include "streams.h"
+//#include "util.h"
+
```
to be removed?

gandrewstone replied:
I want to "additively remove" everything from this file to make it easy to grab new versions.

sickpig replied:
I seed

sickpig commented on xversionmessage.cpp:
```diff
@@ -0,0 +1,62 @@
+        }
+        catch (...)
+        {
+            // LOG(NET, "Error reading extended configuration key %016llx as u64c. Assuming zero.\n", k);
```
to be removed

gandrewstone replied:
additive removal

sickpig commented on xversionmessage.cpp:
```diff
@@ -0,0 +1,62 @@
+    std::vector<uint8_t> vec;
+    vec.insert(vec.begin(), s.begin(), s.end());
+    xmap[key] = vec;
+    // LOCK(cacheProtector);
```
to be removed?

gandrewstone replied:
additive removal

sickpig commented on xversionmessage.h:
```diff
@@ -0,0 +1,125 @@
+        auto nType = s.nType;
+        auto nVersion = s.nVersion;
+        READWRITE(REF(CompactMapSerialization(xmap)));
+        //if (nSerSize > MAX_XVERSION_MAP_SIZE)
```
why this check is commented?

gandrewstone replied:
I don't think that the seeder needs to be enforcing the rules like bitcoind nodes do...

sickpig replied:
I see. 

But doesn't make sense anyway to "signal" that something wrong so that we could avoid adding nodes that are not xversion complaint?

gandrewstone replied:
Yes, so I was thinking that in the worst case we give an IP of a bogus node.  And I was worried that this code doesn't catch that exception.  However, I looked at serialize.h and it turns out that the exception is already being thrown by other incorrect deserialization operations.  So I've added this code back in.

sickpig commented on main.cpp:
```diff
@@ -270,10 +280,11 @@ extern "C" void* ThreadCrawler(void* data)
             bool getaddr = res.ourLastSuccess + 86400 < now;
-            res.fGood = TestNode(res.service, res.nBanTime, res.nClientV, res.strClientV, res.nHeight, getaddr ? &addr : NULL);
+            res.fGood = TestNode(res.service, res.grapheneVersion, res.electrsVersion, res.capdVersion, res.nBanTime, res.nClientV, res.strClientV, res.nHeight, getaddr ? &addr : NULL);
+            //    if (res.fGood) printf("\nClient: %s, graphene: %lu\n", res.strClientV.c_str(), (long unsigned int) res.grapheneVersion);
```
to be removed?

gandrewstone replied:
its a pretty useful printout for debugging

sickpig replied:
I wonder that once we get this done including this source code into the main BU repo would made this kind of scenario more convenient by using LOG/LOGA macros

gandrewstone replied:
yes, we should definitely convert all these ugly printouts to LOG messages when we convert. I don't think its worth it to move the LOG infra over here though.

sickpig commented on main.cpp:
```diff
@@ -371,6 +382,7 @@ extern "C" int GetIPList(void* data, char* requestedHostname, addr_t* addr, int
 
     uint64_t requestedFlags = 0;
     int hostlen = strlen(requestedHostname);
+    // printf("requested hostname: %s\n", requestedHostname);
```
to be removed?

gandrewstone replied:
This code uses commented out printfs all over the place to leave debugging code around that should not be part of the normal output.

sickpig replied:
I agree, the code use a lot of `printf`. It's just that I rather don't add new occurrences. That said we could take care of it once we move everything in the BU codebase.

sickpig commented on xversionmessage.h:
```diff
@@ -0,0 +1,125 @@
+#include "serialize.h"
+// #include "sync.h"
+// #include "tinyformat.h"
+// #include "utilstrencodings.h"
```
to be removed? 

wonder if we could mving all the code inside `src/seeder` in the main BU BCH client repo, so to avoid to copy over here and maintain in sync `xversion*` files

edit: I see you proposed the same in the PR description

gandrewstone replied:
yes.  Doing it would be a large job compared to this PR though, but I think would be easier than cleaning up this entire code base.  So we should consider doing it if we get time to do a cleanup.

sickpig commented on bitcoin.cpp:
```diff
@@ -27,6 +29,12 @@ class CNode
+public:
+    uint64_t maxGrapheneVersion = 0;
+    uint64_t electrsVersion = 0;
+    uint64_t capdVersion = 0;
```
since all others info we gathered are private I wonder why declaring those members as public? Any specific need?

gandrewstone replied:
I mean for these variables to be part of the external interface to this class.  For example, if you want to know the CNode's electrsVersion, read the variable.

sickpig replied:
I see. 

Would it make sense for all the other variables that now are read via accessors e.g. `GetClientVersion()`, `GetClientSubVersion()`, ... ?

sickpig commented on serialize.h:
```diff
@@ -1260,6 +1278,31 @@ class CDataStream
+ * "SerializationOp" template, which implements the body of each class' serialization
+ * code. Adding "ADD_SERIALIZE_METHODS" in the body of the class causes these wrappers to be
+ * added as members.
+ */
```
dumb question probably, but why do we need those serialization methods? I mean we should be able to store/get the needed info for each node properties like we do previous to this PR?

gandrewstone replied:
Yes, I could have ported the Xversion message serialization and deserialization backwards to the old serialization style.  But instead I decided to add the new serialization style and make them both work simultaneously in this project.  This is kind of moving things forwards, even though its only moving the new stuff forwards.

sickpig commented on bitcoin.cpp:
```diff
@@ -158,6 +166,22 @@ class CNode
+            BeginMessage("xverack");
+            EndMessage();
+
+            if (!vAddr)  // Early quit if we get an XVERSION message sine that's the last thing we need if we aren't looking for an ADDR message
```
s/sine/since/ in the inline comment

gandrewstone replied:
done

sickpig commented on bitcoin.cpp:
```diff
@@ -158,6 +166,22 @@ class CNode
+            // printf("XVERSION received\n");
+            CXVersionMessage msg;
+            vRecv >> msg;
+            maxGrapheneVersion = msg.as_u64c(XVer::BU_GRAPHENE_MAX_VERSION_SUPPORTED);
```
I'm never looked at the BU xversion implementation but I guess that if the xversion payload has a key different from `XVer::BU_GRAPHENE_MAX_VERSION_SUPPORTED` then `maxGrapheneVersion` would be equal to 0, am I correct?

gandrewstone replied:
Yes, as_u64c returns 0 if the specified key does not exist.

gandrewstone committed:
add electrumserver detection

gandrewstone committed:
fix PR comments

gandrewstone commented:
I addressed your comments and also added support for reporting electrum protocol servers since dagur enabled that in xversion in BU.  I have 2 servers currently running on mainnet that advertise this.  You can "force" the discovery of one of these servers by using the -s parameter, for example:
dnsseed -h seed.bitcoinunlimited.net -n seed.bitcoinunlimited.net -p 12345 -m email@example.com -s 68.183.223.81

sickpig commented:
> I addressed your comments and also added support for reporting electrum protocol servers since dagur enabled that in xversion in BU. I have 2 servers currently running on mainnet that advertise this. You can "force" the discovery of one of these servers by using the -s parameter, for example:
> dnsseed -h seed.bitcoinunlimited.net -n seed.bitcoinunlimited.net -p 12345 -m [email@example.com](mailto:email@example.com) -s 68.183.223.81


would you mind adding a NS record on bitcoinunlimited.net DNS for the seed subdomain so that I could test it from the "outside"? 

Having said that the code looks good to me, I'll test in one domain of mine and if everything will be ok I 'll merge the PR.

Thanks!

sickpig commented:
tried to compile your branch on my test box and I got this error:

```
g++ -std=c++11 -pthread -O0 -g -march=native -Wall -Wno-unused -Wno-sign-compare -Wno-reorder -Wno-comment -c -o xversionmessage.o xversionmessage.cpp
In file included from xversionmessage.cpp:6:0:
xversionmessage.h: In member function ‘unsigned int CXVersionMessage::SerializationOp(Stream&, Operation)’:
xversionmessage.h:120:101: error: there are no arguments to ‘strprintf’ that depend on a template parameter, so a declaration of ‘strprintf’ must be available [-fpermissive]
                 strprintf("A version message xmap might at most be %d bytes.", MAX_XVERSION_MAP_SIZE));
                                                                                                     ^
xversionmessage.h:120:101: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
xversionmessage.h: In instantiation of ‘unsigned int CXVersionMessage::SerializationOp(Stream&, Operation) [with Stream = CDataStream; Operation = CSerActionSerialize]’:
xversionmessage.h:103:5:   required from here
xversionmessage.h:120:26: error: ‘strprintf’ was not declared in this scope
                 strprintf("A version message xmap might at most be %d bytes.", MAX_XVERSION_MAP_SIZE));
                          ^
Makefile:9: recipe for target 'xversionmessage.o' failed
make: *** [xversionmessage.o] Error 1
```

sickpig commented on xversionmessage.h:
```diff
@@ -0,0 +1,125 @@
+    template <typename Stream, typename Operation>
+    inline unsigned int SerializationOp(Stream &s, Operation ser_action)
+    {
+        unsigned int nSerSize = 0;
```
what's `nSerSize` used for? It is initialize at 0, MAX_VERSION_MAP_SIZE is set to 100000, hence the `then` branch of the the `if` below never got executed. The code in BU is something like that: 

```c++
    inline void SerializationOp(Stream &s, Operation ser_action)
    {
        READWRITE(REF(CompactMapSerialization(xmap)));
        if (GetSerializeSize(REF(CompactMapSerialization(xmap)), SER_NETWORK, PROTOCOL_VERSION) > MAX_XVERSION_MAP_SIZE)
            throw std::ios_base::failure(
                strprintf("A version message xmap might at most be %d bytes.", MAX_XVERSION_MAP_SIZE));
    }
```

which actually enforce the limit of 100KB for the serialized size of the xversion payload.

sickpig reviewed (changes requested):
https://github.com/sickpig/bitcoin-seeder/pull/5#pullrequestreview-327652281

sickpig commented on main.cpp:
```diff
@@ -380,36 +435,61 @@ extern "C" int GetIPList(void* data, char* requestedHostname, addr_t* addr, int
+    {
+        return SelectRandomByXversion(XVer::BU_GRAPHENE_MAX_VERSION_SUPPORTED, addr, max);
+    }
+    else    if ((sz==strlen("electrumserver")) && (strncasecmp(requestedHostname, "electrumserver", sz)==0))
```
4 spaces between `else` and `if`

sickpig commented:
@gandrewstone ping

gandrewstone committed:
useful messages

gandrewstone committed:
add nextchain

gandrewstone committed:
fix compile error

gandrewstone committed:
add ability to inject 1 known address

gandrewstone committed:
add comments on how to run

sickpig commented:
> tried to compile your branch on my test box and I got this error:
> 
> ```
> g++ -std=c++11 -pthread -O0 -g -march=native -Wall -Wno-unused -Wno-sign-compare -Wno-reorder -Wno-comment -c -o xversionmessage.o xversionmessage.cpp
> In file included from xversionmessage.cpp:6:0:
> xversionmessage.h: In member function ‘unsigned int CXVersionMessage::SerializationOp(Stream&, Operation)’:
> xversionmessage.h:120:101: error: there are no arguments to ‘strprintf’ that depend on a template parameter, so a declaration of ‘strprintf’ must be available [-fpermissive]
>                  strprintf("A version message xmap might at most be %d bytes.", MAX_XVERSION_MAP_SIZE));
>                                                                                                      ^
> xversionmessage.h:120:101: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
> xversionmessage.h: In instantiation of ‘unsigned int CXVersionMessage::SerializationOp(Stream&, Operation) [with Stream = CDataStream; Operation = CSerActionSerialize]’:
> xversionmessage.h:103:5:   required from here
> xversionmessage.h:120:26: error: ‘strprintf’ was not declared in this scope
>                  strprintf("A version message xmap might at most be %d bytes.", MAX_XVERSION_MAP_SIZE));
>                           ^
> Makefile:9: recipe for target 'xversionmessage.o' failed
> make: *** [xversionmessage.o] Error 1
> ```

fixed by 786077c

sickpig commented:
> [#5 (review)](https://github.com/sickpig/bitcoin-seeder/pull/5#pullrequestreview-327652281)

fixed by 786077c

gandrewstone committed:
new xversion keys, make more responsive
